pkg/cmd/server/kubernetes/master/controller_volume.go

```diff
@@ -0,0 +1,157 @@
+	}
+
+	eventcast := record.NewBroadcaster()
+	recorder := eventcast.NewRecorder(kapi.Scheme, kclientv1.EventSource{Component: "persistent-volume-binder"})
+	eventcast.StartRecordingToSink(
+		&kv1core.EventSinkImpl{
+			Interface: ctx.ClientBuilder.ClientGoClientOrDie("persistent-volume-binder").CoreV1().Events(""),
+		},
+	)
+
+	plugins, err := probeRecyclableVolumePlugins(ctx.Options.VolumeConfiguration, c.OpenShiftInfrastructureNamespace, c.RecyclerImage, bootstrappolicy.InfraPersistentVolumeRecyclerControllerServiceAccountName)
+	if err != nil {
+		return false, err
+	}
+
+	volumeController := persistentvolumecontroller.NewController(
+		persistentvolumecontroller.ControllerParameters{
+			KubeClient:                ctx.ClientBuilder.ClientOrDie("persistent-volume-binder"),
+			SyncPeriod:                ctx.Options.PVClaimBinderSyncPeriod.Duration,
+			AlphaProvisioner:          alphaProvisioner,
+			VolumePlugins:             plugins,
+			Cloud:                     c.CloudProvider,
+			ClusterName:               ctx.Options.ClusterName,
+			VolumeInformer:            ctx.InformerFactory.Core().V1().PersistentVolumes(),
+			ClaimInformer:             ctx.InformerFactory.Core().V1().PersistentVolumeClaims(),
+			ClassInformer:             ctx.InformerFactory.Storage().V1beta1().StorageClasses(),
+			EventRecorder:             recorder,
+			EnableDynamicProvisioning: ctx.Options.VolumeConfiguration.EnableDynamicProvisioning,
+		})
+
+	go volumeController.Run(ctx.Stop)
+
+	return true, nil
+}
+
+type PersistentVolumeAttachDetachControllerConfig struct {
+	CloudProvider cloudprovider.Interface
+}
+
+func (c *PersistentVolumeAttachDetachControllerConfig) RunController(ctx kubecontroller.ControllerContext) (bool, error) {
+	attachDetachController, err := attachdetachcontroller.NewAttachDetachController(
+		ctx.ClientBuilder.ClientOrDie("attachdetach-controller"),
+		ctx.InformerFactory.Core().V1().Pods(),
+		ctx.InformerFactory.Core().V1().Nodes(),
+		ctx.InformerFactory.Core().V1().PersistentVolumeClaims(),
+		ctx.InformerFactory.Core().V1().PersistentVolumes(),
+		c.CloudProvider,
+		kctrlmgr.ProbeAttachableVolumePlugins(ctx.Options.VolumeConfiguration),
+		ctx.Options.DisableAttachDetachReconcilerSync,
+		ctx.Options.ReconcilerSyncLoopPeriod.Duration,
+	)
+	if err != nil {
+		return false, fmt.Errorf("failed to start attach/detach persistent volume controller: %v", err)
+	}
+
+	go attachDetachController.Run(ctx.Stop)
+
+	return true, nil
+}
+
+// probeRecyclableVolumePlugins collects all persistent volume plugins into an easy to use list.
+// TODO: Move this into some helper package?
+func probeRecyclableVolumePlugins(config componentconfig.VolumeConfiguration, namespace, recyclerImageName, recyclerServiceAccountName string) ([]volume.VolumePlugin, error) {
+	uid := int64(0)
+	defaultScrubPod := volume.NewPersistentVolumeRecyclerPodTemplate()
+	defaultScrubPod.Namespace = namespace
```

Comment:
This is effectively openshift-infra.  I would have expected the recycling pods to run in namespace dedicated for that purpose just to keep pods out of openshift-infra.